ConsumerApp/Program.cs

```diff
@@ -10,25 +10,27 @@ static void Main()
             Console.WriteLine("> Press ENTER to install hooks and use WinRT component.");
             Console.ReadLine();
 
-            if (InstallRoFunctionHooks())
-            {
-                UseWinRtComponent_DefaultCtor();
-                UseWinRtComponent_Factory();
-                UseWinRtComponent_StaticMethod();
-                UseWinRtComponent_Throwing();
+            var managedHooksManager = new ManagedHooksManager.HooksManager();
+            managedHooksManager.SetupHooks();
+            managedHooksManager.RegisterWinRtType("WinRtComponent.NativeClass, WinRtComponent, ContentType=WindowsRuntime");
+            //RegisterWinRtTypes(managedHooksManager);
 
-                Console.WriteLine(Environment.NewLine + "> Now press ENTER to remove hooks and finish.");
-                Console.ReadLine();
+            UseWinRtComponent_DefaultCtor();
+            UseWinRtComponent_Factory();
+            UseWinRtComponent_StaticMethod();
+            UseWinRtComponent_Throwing();
 
-                RemoveRoFunctionHooks();
-            }
-            else
-            {
-                Console.WriteLine("Failed to install hooks. Press ENTER to finish.");
-                Console.ReadLine();
-            }
+            Console.WriteLine();
+            Console.WriteLine("> Now press ENTER to remove hooks and finish.");
+            Console.ReadLine();
         }
 
+        //[MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
In order to register WinRT type using `typeof` or generic method, the type should be known to the JIT by the time it compiles the method in which the registration occurs. This means that `SetupHooks` must be called before JIT will compile such a method.
In this example, the `RegisterWinRtTypes` method could be optimized away and inlined, thus rendering `SetupHooks` call impossible to call before JIT compiles the registration code. That's why `[MethodImpl(NoInlining)]` attribute is applied here.
The other way to register the type is to use an overload that accepts string. In this case, your code won't need to know about WinRT type for compilation, and you won't need to keep JIT in mind.